packages/list-store/src/use-searchable-facet.tsx

```diff
@@ -0,0 +1,227 @@
+
+export type FacetSortBy = 'alphabetical' | 'count';
+
+export interface Filter {
+  name?: string | number;
+  id: string | number;
+  totalCount: number;
+}
+
+export interface SearchableFilter extends Filter {
+  letterCategory: string;
+  // In order to make the filter searchable, we need to convert the name to a string
+  name: string;
+}
+
+interface FacetProps {
+  searchValue: string;
+  letterCategory: string;
+  sortBy: FacetSortBy;
+  filters: Filter[];
+}
+
+interface FacetState extends FacetProps {
+  setSearchValue: (value: string) => void;
+  toggleLetterCategory: (value: string) => void;
+  setSortBy: (value: FacetSortBy) => void;
+  setFilters: (value: Filter[]) => void;
+}
+
+type FacetStore = ReturnType<typeof createFacetStore>;
+
+export const initialState = {
+  searchValue: '',
+  letterCategory: '',
+  sortBy: 'count' as FacetSortBy,
+  filters: [],
+};
+
+export const createFacetStore = (initProps?: Partial<FacetProps>) => {
+  return createStore<FacetState>()(set => ({
+    ...initialState,
+    ...initProps,
+    setSearchValue: (searchValue: string) =>
+      set(() => ({
+        searchValue,
+      })),
+    toggleLetterCategory: (category: string) =>
+      set(state => ({
+        letterCategory: state.letterCategory === category ? '' : category,
+      })),
+    setSortBy: (sortBy: FacetSortBy) =>
+      set(() => ({
+        sortBy,
+      })),
+    setFilters: (filters: Filter[]) =>
+      set(() => ({
+        filters,
+      })),
+  }));
+};
+
+type FacetProviderProps = React.PropsWithChildren<Partial<FacetProps>>;
+
+export const FacetContext = createContext<FacetStore | null>(null);
+
+export function FacetProvider({children, ...props}: FacetProviderProps) {
+  const storeRef = useRef<FacetStore>();
+  if (!storeRef.current) {
+    storeRef.current = createFacetStore(props);
+  }
+  return (
+    <FacetContext.Provider value={storeRef.current}>
+      {children}
+    </FacetContext.Provider>
+  );
+}
+
+interface GetLetterCategoriesProps {
+  filters: SearchableFilter[];
+  letterCategory?: string;
+  searchValue?: string;
+}
+
+export function getLetterCategories({
+  filters,
+  letterCategory,
+  searchValue,
+}: GetLetterCategoriesProps) {
+  const filtersFilteredBySearchValue = filterBySearchValue(
+    filters,
+    searchValue
+  );
+
+  return [
+    ...new Set([
+      // Always include the letter category of the current filter
+      ...(letterCategory ? [letterCategory] : []),
+      ...filtersFilteredBySearchValue
+        .map(filter => filter.letterCategory)
+        .filter(category => !!category),
+    ]),
+  ].sort();
+}
+
+export function extendFiltersWithLetterCategory(filters?: Filter[]) {
+  if (!filters?.length) {
+    return [];
+  }
+  return filters.map(filter => {
+    let letterCategory = '';
+    const name = filter.name?.toString();
+    if (name) {
+      const firstLetter = name.toUpperCase().replace(/[^0-9A-Z]/g, '')[0];
```

Comment:
1. Very minor: instead of removing _all_ chars that don't meet the criteria, you could just remove the _leading_ chars; makes the regex execution perhaps more efficient.
2. Curious: can `name` be empty? If so, what happens if you pick the first char with `[0]`?

Reply:
1: 👍 

2: The `if (name) {` on line 116 prevents empty names. But `name.toUpperCase().replace(/[^0-9A-Z]/g, '')` could still result in an empty string. `''[0]` will result in `undefined`.

If `letterCategory` is `undefined`, it would be ignored by the function `getLetterCategories`. So filters with the name `undefined`, "", "$%@" cannot be found in the letter category filter, I think this is ok.

I have updated the code a little bit. `letterCategory` is now optional, and I have added tests with the names `undefined`, "" and "#".